renderer/BufferObject.cpp

```diff
@@ -0,0 +1,221 @@
+		__m128i d5 = _mm_load_si128( ( __m128i * )&src[i + 5 * 16] );
+		__m128i d6 = _mm_load_si128( ( __m128i * )&src[i + 6 * 16] );
+		__m128i d7 = _mm_load_si128( ( __m128i * )&src[i + 7 * 16] );
+		_mm_stream_si128( ( __m128i * )&dst[i + 0 * 16], d0 );
+		_mm_stream_si128( ( __m128i * )&dst[i + 1 * 16], d1 );
+		_mm_stream_si128( ( __m128i * )&dst[i + 2 * 16], d2 );
+		_mm_stream_si128( ( __m128i * )&dst[i + 3 * 16], d3 );
+		_mm_stream_si128( ( __m128i * )&dst[i + 4 * 16], d4 );
+		_mm_stream_si128( ( __m128i * )&dst[i + 5 * 16], d5 );
+		_mm_stream_si128( ( __m128i * )&dst[i + 6 * 16], d6 );
+		_mm_stream_si128( ( __m128i * )&dst[i + 7 * 16], d7 );
+	}
+	for( ; i + 16 <= numBytes; i += 16 ) {
+		__m128i d = _mm_load_si128( ( __m128i * )&src[i] );
+		_mm_stream_si128( ( __m128i * )&dst[i], d );
+	}
+	for( ; i + 4 <= numBytes; i += 4 ) {
+		*(uint32 *)&dst[i] = *(const uint32 *)&src[i];
+	}
+	for( ; i < numBytes; i++ ) {
+		dst[i] = src[i];
+	}
+	_mm_sfence();
+}
+
+#else
+
+void CopyBuffer( byte *dst, const byte *src, int numBytes ) {
+	assert_16_byte_aligned( dst );
+	assert_16_byte_aligned( src );
+	SIMDProcessor->Memcpy( dst, src, numBytes );
+}
+
+#endif
+
+/*
+========================
+BufferObject
+========================
+*/
+BufferObject::BufferObject( GLenum targetType, GLenum usage ) {
+	size = 0;
+	bufferObject = 0;
+	bufferType = targetType;
+	bufferUsage = usage;
+	SetUnmapped();
+}
+
+/*
+========================
+~BufferObject
+========================
+*/
+BufferObject::~BufferObject() {
+	FreeBufferObject();
+}
+
+/*
+========================
+AllocBufferObject
+========================
+*/
+void BufferObject::AllocBufferObject( int allocSize ) {
+	assert( bufferObject == 0 );
+	assert_16_byte_aligned( data );
+
+	if( allocSize <= 0 ) {
+		idLib::Error( "BufferObject::AllocBufferObject: allocSize = %i", allocSize );
+	}
+
+	size = allocSize;
+	int numBytes = GetAllocedSize();
+
+	// clear out any previous error
+	qglGetError();
+
+	qglGenBuffersARB( 1, &bufferObject );
+	if( bufferObject == 0 ) {
+		common->FatalError( "BufferObject::AllocBufferObject: failed" );
+	}
+	qglBindBufferARB( bufferType, bufferObject );
+	qglBufferDataARB( bufferType, numBytes, nullptr, bufferUsage );
+
+	GLenum err = qglGetError();
+	if( err == GL_OUT_OF_MEMORY ) {
+		common->FatalError( "BufferObject::AllocBufferObject: allocation failed - out of memory" );
+	}
+}
+
+/*
+========================
+FreeBufferObject
+========================
+*/
+void BufferObject::FreeBufferObject() {
+	if( bufferObject == 0 ) {
+		return;
+	}
+
+	if( IsMapped() ) {
+		UnmapBuffer();
+	}
+
+	qglDeleteBuffersARB( 1, &bufferObject );
+
+	size = 0;
+	bufferObject = 0;
+}
+
+/*
+========================
+MapBuffer
+========================
+*/
+void * BufferObject::MapBuffer( int mapOffset ) {
+	assert( bufferObject != 0 );
+	assert( IsMapped() == false );
+
+	void *buffer = NULL;
+
+	qglBindBufferARB( bufferType, bufferObject );
+	// orphan old contents
```

Comment:
We need to draw some VBO contents, and in the process fill another VBO, which would be drawn later.
As far as I understand, orphaning VBO is one way to achieve it (orphaned data is being read from, while you can fill the current VBO data). Another approach is to have several (e.g. two) VBOs and and switch between them: one is filled, another is being drawn from.
It seems from the code that both ways are used. Not that I understand well what's going on, but isn't it a sign that something is wrong?...

Reply:
Your last point would require persistent mapping, which is an OpenGL 4.2 feature :)

But the problem here is that frontend and backend work in parallel. So if backend is rendering from buffer A and frontend is writing to buffer B, that's fine. But if you now switch the buffers and have the frontend write to buffer A without orphaning, there is a good chance that graphics card is still rendering from it. OpenGL is asynchronous, after all. So if you don't want to orphan buffer A, you'll have to wait until all relevant draw calls are truly finished, which causes stalls. Alternatively, we could use triple buffers with syncing, where stalls should virtually never happen. But they do require an additional buffer.

Reply:
So you want to say that three buffers are needed to avoid stalls. That's because frontend is filling buffer for the next frame, backend is sending draw commands for the current frame, and GPU is rendering data for the previous frame, right?
I don't see orphaning in BFG code. I think explicit double/triple buffering gives more control over how much RAM is used (orphaning makes it unlimited/driver-dependent). Also, is it much less dependent on driver implementation (with orphaning you hope that driver supports the concept and has enough).
For example, isn't it true that the current way with orphanin uses at least 4 buffers? There are two VBOs, which are often swapped, and each of them regularly gets an orphaned version also. Does GPU allocate/deallocate those orphaned arrays immediately, or it simply increases number of orphaned versions until no one wants more?

Reply:
Well, BFG says GL_MAP_INVALIDATE_RANGE_BIT when mapping which supposedly should also orphan the range when needed, but it seems to not work correctly with some drivers. The reason why BFG gets away with it, I think, is because they do VSync a little differently and wait for it at a different place, which functions essentially like a sync. But I'm just guessing, here.

Since HMart reported performance drops with the current explicit glBufferData call, I've already switched to a triple-buffered approach with sync, just haven't pushed yet.
Downside is that the required glFenceSync is an OpenGL 3.2 feature, or an arb extension. But I think any reasonable driver today should probably support it.

Reply:
ARB_sync depends on OpenGL 3.1, so it is safe to assume that anything without OpenGL 3.1 will stop working.
I wonder if it would be possible to make it optional. Maybe we can insert some stupid glFinish if ARC_sync is missing? If it is possible and won't be horrendously slow, and it will allow for older OpenGL implementations to work at least somehow.

Reply:
We can do that. But even BFG required OpenGL 3 core profile. I wonder if there are really any GPUs out there that are not compatible with OpenGL 3.1 and are able to play TDM otherwise?

Reply:
Well, I know one case: Chromium-based driver in VirtualBox's Linux VM.
I have used it a lot for reviving linux build and even fixing bugs/crashes in it, despite the fact that it is barely playable. If this stops working completely, then TDM will be left with zero Linux-side developers again =(
I installed real Linux some years ago, and my experience shows that I never used it. In case of VM, I can even use Visual Studio for code browsing, combined with gdb-based debuggers =)
I understand that this stupid case does not justify anything...

Reply:
Hm, fair point. But if performance doesn't matter, couldn't it potentially work with a Mesa software driver? I think that one is fairly advanced feature-wise.
In any case, I implemented glFenceSync for now without fallback, just to see if it works as intended. I'll keep this in mind once I've addressed any other issues.

Reply:
Finally found it. BFG also uses glFenceSync, but they do it immediately after SwapBuffers. This is why they get away with 2 buffers without explicit orphaning. They do use glFinish as a fallback.
I think triple-buffering is a cleaner solution, since OpenGL can definitely queue more than one frame, so syncing on SwapBuffers could still cause stalls.

Reply:
I added the fallback to glFinish.